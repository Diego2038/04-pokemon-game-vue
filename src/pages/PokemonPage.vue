<template>
  <div v-if="!pokemon">
    <h1>Por favor espere...</h1>
  </div>
  <div v-else class="fade-in">
    <h1>¿Quién es este Pokemon?</h1>
    <PokemonPicture :pokemonId=" pokemon.id " :showPokemon="showPokemon"></PokemonPicture>
    <PokemonOptions :pokemonsArray="pokemonArr"
      @selection="checkAnswer"> <!-- También lo puedes poner sin el parámetro, pero lo pongo para que sepas lo que significa-->
    </PokemonOptions>
    <!-- OJO: Habías probado trasladar los datos por medio de una función, pero las promesas no se resolvían. En cambio, cuando los trasladabas con una data, tiene todo el tiempo del mundo para esperar y actualizarse, por lo tanto con ello sí se actualizaba de manera idónea-->

    <template v-if="showAnswer" class="fade-in">
      <h3> {{message}}</h3>
      <button @click="newGame">
        Nuevo juego
      </button>
    </template>

  </div>
  
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions.vue'
import PokemonPicture from '@/components/PokemonPicture.vue'
import getPokemonsOptions from '@/helpers/getPokemonOptions'


export default {
  name : 'pokemon-component',
  components : {
    PokemonOptions,
    PokemonPicture
  },
  data() {
    return {
      pokemonArr : [],
      pokemon : null,
      showPokemon : false,
      message : '',
      showAnswer : false
    }
  },
  methods : {
    async getPokemonsMixed() {
      // console.log(getPokemonsOptions())
      // return await getPokemonsOptions()
      this.pokemonArr = await getPokemonsOptions()
      // console.log(this.pokemonArr)
      const randInt = Math.floor( Math.random() * 4  )
      
      this.pokemon = this.pokemonArr[ randInt ]
    },

    checkAnswer({pokemon}) { // pokemonxD es lo que retorna el evento, definido en el componente hijo (y en el primer parámetro si llegaras a ponerlo)
      // console.log('Pokemon page llamado', pokemonxD.name, pokemonxD.id)
      // console.log( pokemonxD, otro, otroMas) 

      if( this.pokemon.id !== pokemon.id ){
        this.message = 'No es el pokemon correcto'
      } else {
        this.message = `Felicidades, acertaste, es ${pokemon.name}`
        this.showPokemon = true
      }
      this.showAnswer = true

      console.log('visto', pokemon.name, this.showPokemon )
    },

    async newGame () {
      this.pokemonArr = [],
      this.pokemon = null,
      this.showPokemon = false,
      this.message = '',
      this.showAnswer = false 
      await this.getPokemonsMixed()

      // Agregué un llamado retartado debido a que, por algún motivo, al hacer el llamado a la API pokemon no se actualizaba. Correción: El error estaba en Math.round, se cambió a Math.floor para que no genere un número 4 para un índice de sólo 4 objetos
      setTimeout(async () => {
        if ( !this.pokemon ){
          await this.getPokemonsMixed()
        }
      }, 2500)
      
      // this.updateComponente()
    },

    // updateComponente() {
    //   setTimeout(() => {
    //     if( !!pokemon) {
    //       this.newGame()
    //     }
    //   },1000)
      
    // }
  },
  mounted() {
    this.getPokemonsMixed()
  }
}
</script>

<style>

</style>