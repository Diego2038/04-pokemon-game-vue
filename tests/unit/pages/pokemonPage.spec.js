import { shallowMount, mount } from "@vue/test-utils";
import  PokemonPage  from "@/pages/PokemonPage";
import  {pokemons } from "../mocks/pokemons.mock"

describe('PokemonPage Component', () => {

  let wrapper
  let getPokemonsMixedSpy
  beforeEach(() => {
    getPokemonsMixedSpy = jest.spyOn( PokemonPage.methods, 'getPokemonsMixed')
    
    wrapper = shallowMount( PokemonPage )

    // jest.clearAllMocks() // Para reiniciar todos los componentes de las pruebas, y así no hayan inconsistencias. OJO, pero si inicializas el spy en beforeEach, no podrás limpiar todos los mocks
  })
  
  test('Debe de hacer match con la snapshot', () => {
    console.log( PokemonPage.methods )
    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('Debe de llamar mixPokemonArray al montarse el componente', () => {
    // Se debe de insertar el espía antes que se dispare el evento para que pueda escucharlo, sino, no podrá escucharlo cuando se dispare. Por eso se puso al comienzo antes del wrapper
    expect( getPokemonsMixedSpy ).toHaveBeenCalled()
  })

  test('Debe de hacer match con el Snapshot cuando cargan todos los pokemones', () => {
    // OJO: La diferencia entre el shallowMount y el mount, es que el shallowMount va a montar el componente sútilmente, simplificando mucha de sus partes y obviando otras para que sea lo más ligero posible. Mientras que el mount lo levanta como si estuviera montado en verdad, ejecutando como si fuera producción. Desde luego ambos son excelentes dependiendo del contexto
    const wrapper = shallowMount( PokemonPage, {
      data() {
        return {
          pokemonArr : pokemons,
          pokemon : pokemons[2],
          showPokemon : false,
          message : '',
          showAnswer : false
        }
      }
    })

    // console.log( wrapper.html())
    expect( wrapper.html()).toMatchSnapshot()

  })

  test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

    const wrapper = shallowMount( PokemonPage, {
      data() {
        return {
          pokemonArr : pokemons,
          pokemon : pokemons[2], // id = 3
          showPokemon : false,
          message : '',
          showAnswer : false
        }
      }
    })

    // NOTA: con shallowMounted dichos componentes están "hechos a medias", y sus nombres tienen agregados 'stub' y separado con -
    const pokemonPicture = wrapper.find('pokemon-picture-stub') 
    const pokemonOptions = wrapper.find('pokemon-options-stub') 

    // PokemonPicture debe de existir 
    expect( pokemonPicture.exists() ).toBeTruthy()
    // PokemonOptions debe de existir
    expect( pokemonOptions.exists() ).toBeTruthy()

    // Pokemon picture attribute pokemonId === 3
    // console.log( wrapper.html()) // Vital para saber cómo hacer estas especificaciones
    expect( pokemonPicture.attributes('pokemonid')).toBe("3")

    // PokemonOptions attribute pokemons toBe true
    // console.log( pokemonOptions.attributes('pokemonsarray') )
    // expect( pokemonOptions.attributes('pokemonsarray') ).not.toEqual(undefined) 
    expect( pokemonOptions.attributes('pokemonsarray') ).toBeTruthy() 

  })

  test('pruebas con checkAnswer', async () => {
    const wrapper = shallowMount( PokemonPage, {
      data() {
        return {
          pokemonArr : pokemons,
          pokemon : pokemons[2], // id = 3
          showPokemon : false,
          message : '',
          showAnswer : false
        }
      }
    }) 

    await wrapper.vm.checkAnswer( {pokemon: pokemons[2] , uno: 'uno', dos: 'dos'} )
    // console.log( wrapper.html() )

    expect( wrapper.find('h3').exists()).toBeTruthy()

    expect( wrapper.vm.showPokemon ).toBeTruthy()

    expect( wrapper.find('h3').text()).toBe(`Felicidades, acertaste, es ${pokemons[2].name}`)

    await wrapper.vm.checkAnswer( {pokemon: pokemons[0]} )

    expect( wrapper.vm.message).toBe('No es el pokemon correcto')

  })




})