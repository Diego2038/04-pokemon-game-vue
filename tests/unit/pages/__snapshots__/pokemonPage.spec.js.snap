// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe de hacer match con el Snapshot cuando cargan todos los pokemones 1`] = `
<div class="fade-in">
  <h1>¿Quién es este Pokemon?</h1>
  <pokemon-picture-stub pokemonid="3" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemonsarray="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub><!-- OJO: Habías probado trasladar los datos por medio de una función, pero las promesas no se resolvían. En cambio, cuando los trasladabas con una data, tiene todo el tiempo del mundo para esperar y actualizarse, por lo tanto con ello sí se actualizaba de manera idónea-->
  <!--v-if-->
</div>
`;

exports[`PokemonPage Component Debe de hacer match con la snapshot 1`] = `
<div>
  <h1>Por favor espere...</h1>
</div>
`;
